### ISSUE TITLE
MeanSquaredError truncated to 32-bit precision when using 64-bit 

### ISSUE THREAD

**Issue by jonwittmer** [keras-team/keras#15791]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab
- TensorFlow installed from (source or binary): Colab
- TensorFlow version (use command below): 2.7.0
- Python version: 3.7.12

Problem Description: This is a re-opening of the issue [here](https://github.com/tensorflow/tensorflow/issues/39878) which was closed due to inactivity. I was directed to reopen the issue here. A short summary is that it seems some functions seem to internally convert tensors to float32, regardless of whether the input tensor is float32 or float64. This is evidenced in MeanSquaredError and by the following code. One would expect that calling a function with float64 tensors would maintain 64 bit precision throughout the computation, but the computation is only being executed with 32 bit precision. I just tested with Tensorflow 2.7.0 and the issue remains. This is a problem because it gives the user unfounded confidence in the precision of training with 64 bit variables when in reality, only 32 bit precision is achieved. Either float64 tensors need to be fully supported, not supported at all, or a message clearly indicating the limitation of the Keras functions in working with float64 tensors. 

- Do you want to contribute a PR? (yes/no): not at this time

**Standalone code to reproduce the issue**.

```
import tensorflow as tf
from tensorflow import keras
import sys

print(sys.version)
print(tf.__version__)

# some sample data 
x = tf.convert_to_tensor([[2.1]], dtype=tf.float64)
y_true = tf.square(x)

# simulated output of NN - something not the same as y_true
y_pred = tf.convert_to_tensor([[3.68]], dtype=tf.float64)

# TF loss
loss = keras.losses.MeanSquaredError()
tf_loss = loss(y_pred, y_true)

# manually computed loss in 64-bit
man_loss64 = tf.square(y_pred - y_true)

# manually computed loss, cast to 32 bit, then back to 64
man_loss32 = tf.cast(tf.cast(tf.square(y_pred - y_true), dtype=tf.float32), dtype=tf.float64)

# difference between loss computations
diff64 = abs(man_loss64 - tf_loss)
diff32 = abs(man_loss32 - tf_loss)
print(f'Difference in 64 bit: {diff64}')
print(f'Difference in 32 bit: {diff32}')
```

**Comment by sanatmpa1** [comment]
I am able to reproduce the issue reported in `Colab` running with `TF 2.7.0`. Please find the [gist here](https://colab.sandbox.google.com/gist/sanatmpa1/b2c61ec4af5f67a7bfd30f56bec65b11/15791.ipynb)

**Comment by LukeWood** [comment]
Notes from triage:

It looks like we may be truncating the values somewhere internally on accident, not immediately obvious where this is occurring.

**Comment by qlzh727** [comment]
Thanks for the reporting. Let me take a closer look

**Comment by qlzh727** [comment]
I think this might related to the autograph which we use for loss function conversion. https://github.com/keras-team/keras/blob/a17ead8ae150869ecdf733a0a6a04695a7055517/keras/losses.py#L244. 

When I change the loss in example above to keras.losses.mean_sequared_error, the diff goes away. @mdanatg for more inputs about autograph.

**Comment by mdanatg** [comment]
@qlzh727 autograph doesn't make any precision-changing transformations, so I doubt it's that. The layer version has additional logic in the parent class' `Layers.__call__`, such as weighting - I suggest having a closer look there.

**Comment by qlzh727** [comment]
@mdanatg, thanks for the reply. With closer look, I think the issue is actually from https://github.com/keras-team/keras/blob/a17ead8ae150869ecdf733a0a6a04695a7055517/keras/losses.py#L142, and https://github.com/keras-team/keras/blob/a17ead8ae150869ecdf733a0a6a04695a7055517/keras/utils/losses_utils.py#L318.

let me see if I can fix this on keras side.
